File: src/main/java/codetest/SockMerchant.java
package codetest;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;
import java.util.stream.Stream;

import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;

public class SockMerchant {
    // Complete the sockMerchant function below.
    static int sockMerchant(int n, int[] ar) {
        return Stream.of(ar)
                .collect(groupingBy(i -> i, counting()))
                .entrySet()
                .stream()
                .filter(i -> i.getValue() >= 2)
                .mapToInt(i -> (int) (i.getValue() / 2))
                .sum();
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
//        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        int[] ar = new int[n];

        String[] arItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < n; i++) {
            int arItem = Integer.parseInt(arItems[i]);
            ar[i] = arItem;
        }

        int result = sockMerchant(n, ar);
        System.out.println("result = " + result);

//        bufferedWriter.write(String.valueOf(result));
//        bufferedWriter.newLine();
//
//        bufferedWriter.close();

        scanner.close();
    }
}

File: src/main/kotlin/codetest/hackerrank/SockMerchant.kt
package codetest.hackerrank

import java.util.*

fun sockMerchant(n: Int, ar: Array<Int>): Int {
    return ar.groupBy { it }
            .filter { it.value.size >= 2 }
            .map { it.value.size / 2 }
            .sum()
}

fun main(args: Array<String>) {
    val scan = Scanner(System.`in`)

//    val n = scan.nextLine().trim().toInt()
//    val ar = scan.nextLine().trim().split(" ").map { it.trim().toInt() }.toTypedArray()

    val input = "4 5 5 5 6 6 4 1 4 4 3 6 6 3 6 1 4 5 5 5 "
    val inArr = input.trim().split(" ")

    val n = inArr.size
    val ar = inArr.map { it.trim().toInt() }.toTypedArray()
    val result = sockMerchant(n, ar)

    println(result)
}
